### PR TITLE
Bug Fix- list of input types

### PR DESCRIPTION
**Pull request by yarinvak**
fixes the bug of unsupported list of input types as an argument
fixing issue https://github.com/graphql-java/graphql-java-annotations/issues/78
fixing issue https://github.com/graphql-java/graphql-java-annotations/issues/90

### CODE REVIEW

File: src/main/java/graphql/annotations/GraphQLAnnotations.java [hunk 1]
```diff
@@ -616,6 +616,26 @@ protected GraphQLFieldDefinition getField(Method method) throws GraphQLAnnotatio
...
+            return getWrappedTypesStack(((GraphQLList) graphQLType).getWrappedType(), stack);
+        }
+        else{
+            return stack;
+        }
+    }
+
+    private graphql.schema.GraphQLType getOrPutInRegistry(graphql.schema.GraphQLType graphQLType) {
+        graphql.schema.GraphQLType typeFromRegistry = typeRegistry.get(graphQLType.getName());
+        if (typeFromRegistry!=null)
```
Comment (guy120494) [inline src/main/java/graphql/annotations/GraphQLAnnotations.java]: please use braces in the if and else

File: src/main/java/graphql/annotations/GraphQLAnnotations.java [hunk 2]
```diff
@@ -554,8 +538,24 @@ protected GraphQLFieldDefinition getField(Method method) throws GraphQLAnnotatio
                     graphql.schema.GraphQLType graphQLType = finalTypeFunction.buildType(t, parameter.getAnnotatedType());
                     if (graphQLType instanceof GraphQLObjectType) {
                         GraphQLInputObjectType inputObject = getInputObject((GraphQLObjectType) graphQLType, "input");
+                        inputObject = (GraphQLInputObjectType) getOrPutInRegistry(inputObject);
                         graphQLType = inputObject;
+                    } else if (graphQLType instanceof GraphQLList) {
+                        Stack<graphql.schema.GraphQLType> typesStack= new Stack<>();
+                        typesStack = getWrappedTypesStack(graphQLType, typesStack);
+                        graphql.schema.GraphQLType wrappedType = typesStack.pop();
+                        if (wrappedType instanceof GraphQLObjectType){
```
Comment (guy120494) [inline src/main/java/graphql/annotations/GraphQLAnnotations.java]: What if I have an interface? will this still work? (I don't remember if an interface is also an objectType)

Comment (yarinvak) [reply src/main/java/graphql/annotations/GraphQLAnnotations.java]: It won't, but Im not sure if we should handle it the same way

**Comment by guy120494** [comment]
@tdraier claims that he solved these problems in #94. I think it is better to code review and merge that PR because it fixes other problems as well

**Comment by yarinvak** [comment]
He doesn't solve issue #78 entirely. This PR allows List of List of List (...) of Input Type (it handles it recursively), while his PR does not.
This PR also fixes tests that did not pass in the master branch.
And it also does not change the code drastically compared to the PR you referenced.

**Comment by tdraier** [comment]
Actually List of List of List of Input types work in #94 , I've just added a unit test. It also allows you to use constructor with typed fields instead of simple map, which makes possible to use more complex type ( in the unit test, an input containing a list of other objects )

File: src/main/java/graphql/annotations/GraphQLAnnotations.java [hunk 3]
```diff
@@ -640,6 +648,26 @@ public GraphQLInputObjectType getInputObject(GraphQLObjectType graphQLType, Stri
                         collect(Collectors.toList()));
     }
 
+    private GraphQLInputType handleGraphQLList(graphql.schema.GraphQLType type) {
```
Comment (DoctorVoid) [inline src/main/java/graphql/annotations/GraphQLAnnotations.java:651]: I think it is better that the function get a GraphQLList as an input

**Comment by DoctorVoid** [comment]
Do you have tests that check list of inputs?

File: src/main/java/graphql/annotations/GraphQLAnnotations.java [hunk 4]
```diff
@@ -616,6 +602,26 @@ protected GraphQLFieldDefinition getField(Method method) throws GraphQLAnnotatio
...
+    private Stack<graphql.schema.GraphQLType> getWrappedTypesStack(graphql.schema.GraphQLType graphQLType, Stack<graphql.schema.GraphQLType> stack) {
+        stack.push(graphQLType);
+        if (graphQLType instanceof GraphQLList) {
+            return getWrappedTypesStack(((GraphQLList) graphQLType).getWrappedType(), stack);
+        } else {
+            return stack;
+        }
+    }
+
+    private graphql.schema.GraphQLType getOrPutInRegistry(graphql.schema.GraphQLType graphQLType) {
```
Comment (DoctorVoid) [inline src/main/java/graphql/annotations/GraphQLAnnotations.java:615]: rename method

File: src/main/java/graphql/annotations/GraphQLAnnotations.java [hunk 5]
```diff
@@ -640,6 +648,26 @@ public GraphQLInputObjectType getInputObject(GraphQLObjectType graphQLType, Stri
...
+    private GraphQLInputType handleGraphQLList(graphql.schema.GraphQLType type) {
+        GraphQLInputType inputType;
+        Stack<graphql.schema.GraphQLType> typesStack = new Stack<>();
+        typesStack = getWrappedTypesStack(type, typesStack);
+        graphql.schema.GraphQLType wrappedType = typesStack.pop();
+        if (wrappedType instanceof GraphQLObjectType) {
+            GraphQLInputObjectType inputObject = getInputObject((GraphQLObjectType) wrappedType, "input");
+            inputObject = (GraphQLInputObjectType) getOrPutInRegistry(inputObject);
+            inputType = inputObject;
+            while (!typesStack.isEmpty()) {
```
Comment (DoctorVoid) [inline src/main/java/graphql/annotations/GraphQLAnnotations.java:660]: extract to different method

**Comment by guy120494** [comment]
We just merged @tdraier PR (#94 ), so I'm closing this PR

**Pull request closed without merging**